base/sparse/sparsevector.jl

```diff
@@ -786,13 +786,15 @@ getindex(x::AbstractSparseVector, I::AbstractVector{Bool}) = x[find(I)]
 getindex(x::AbstractSparseVector, I::AbstractArray{Bool}) = x[find(I)]
 @inline function getindex(x::AbstractSparseVector, I::AbstractVector)
     # SparseMatrixCSC has a nicely optimized routine for this; punt
-    S = SparseMatrixCSC(x.n, 1, [1,length(x.nzind)+1], x.nzind, x.nzval)
+    Ti = eltype(x.nzind)
+    S = SparseMatrixCSC(x.n, 1, Ti[1,length(x.nzind)+1], x.nzind, x.nzval)
     S[I, 1]
 end
 
 function getindex(x::AbstractSparseVector, I::AbstractArray)
     # punt to SparseMatrixCSC
-    S = SparseMatrixCSC(x.n, 1, [1,length(x.nzind)+1], x.nzind, x.nzval)
+    Ti = eltype(x.nzind)
+    S = SparseMatrixCSC(x.n, 1, Ti[1,length(x.nzind)+1], x.nzind, x.nzval)
```

Comment:
Would be good with a test case for this change too.

Reply:
Definitely a good idea to have another test there. I added one and found another bug. Here's an example
```
julia> a = sparsevec(Int32[1;4;7],Int32[2;2;2],10)
10-element SparseVector{Int32,Int32} with 3 stored entries:
  [1 ]  =  2
  [4 ]  =  2
  [7 ]  =  2

julia> I = rand(1:10,3,1)
3×1 Array{Int64,2}:
 1
 4
 5

julia> r = a[I]
3×1 SparseMatrixCSC{Int32,Int64} with 2 stored entries:
  [1, 1]  =  2
  [2, 1]  =  2
```
In that example `r` should be a 3 X 1 `SparseMatrixCSC{Int32,Int32}`. In general, calling `A[I]` with `A` a sparse matrix and `I<:AbstractArray` (or A a sparse vector and `I` an abstract array with `ndims(I)>1`) always returns a `SparseMatrixCSC{Tv,Int}` regardless of the indexing integer type of `A`. This arises due to a bug in the sparse matrix getindex method at https://github.com/JuliaLang/julia/blob/caa7f04da63bb13fb56560843ad84b09b33d6280/base/sparse/sparsematrix.jl#L2302.

At least I assume it's a bug. I don't understand why it would be desirable to change the type of the indexing arrays of a slice by default.
 In any event, it's simple to fix this. Is it ok to add that fix to this PR or should I create a another one since the change affects sparse matrices as well as sparse vectors?

Reply:
You can add it to the same PR if you want. Perhaps @Sacha0 can answer the other questions regarding why we always return with `Int`?

Reply:
The index-type non-preservation in the linked methods looks like an oversight on this end as well :).